### PR TITLE
Reimplement forge blockstates variants as pseudo-models (fixes custom model loading for blocks)

### PR DESCRIPTION
**Pull request by gigaherz**
~~aka. HACKS! DIRTY HACKS!~~
~~But it gets the forge blockstates variants to actually load...~~
~~Needs more work, so draft only for now.~~

~~EDIT: Big note, in case anyone gets the wrong idea:~~
~~Yes the patches are dirty and I allowed things like new imports to be added to them. I don't care at this point. If this ever becomes ready for merging, everything will be cleaned up.~~

This is still mostly a workaround, but I believe a sane one. Opinions welcome, testing needed, review wanted.

See my post [down there](https://github.com/MinecraftForge/MinecraftForge/pull/6154#issuecomment-533837764) for more details.

### CODE REVIEW

**Comment by LexManos** [comment]
With 1) Data Generators and 2) The submodel system in vanilla.
Is ForgeBlockState even required anymore?
What features are needed?
Is the permutation generators needed? {Pretty sure this is replaced by submodel system in vanilla}

I would be very happy to have this Forge variants die, if we have valid system to make modder's lives easier using vanilla/simpler mechanics.


Also, thanks for volunteering to write data generators for this :)

**Comment by gigaherz** [comment]
Forge variants are still needed for custom model loading, until we get an alternative for it. This is meant to become a "minimal hack" (not minimal yet) to get the old forge blockstates system working until a new system is in place.

The progression I see for the model system would be like:
step 1: give people something that works for loading custom models (this PR, at least for blocks)
step 2: work on the new model system
step 3: update the hacks from this PR to wrap a "new model" in place of an old model, and deprecate the old model hierarchy
step 4: sometime during early 1.15 dev work, get rid of the forge blockstates loader and related systems, and leave only the new API

**Comment by mcenderdragon** [comment]
Forge Variants are very handy for texture overwrites when you dont want to create 64 model json which are almost the same: https://gist.github.com/mcenderdragon/7c2b63e49a1917789110f6144612e59f.

**Comment by MoreThanHidden** [comment]
Without the Forge blockstate variants or ModelLoader not working I have come accross the following issues:

- **Obj models currently don't load from the vanilla json files** #6109
(**Current workaround:** Manually bake these at ModelBakeEvent and addSprite on TextureStitchEvent.Pre)
- **No way of making multipart item models**
(**Current workaround:** Make a model specifically for use as item or Bake Model)
- **Texture overwrites based on blockstate**
(**Current workaround:** Separate Models for each Texture Variant)
- **Some of the rotations from Forge Blockstate aren't possible with vanilla multipart**
(**Current workaround:** Create a duplicate model rotating the z axis)

**Comment by LexManos** [comment]
The ONLY original intention of ForgeBlockState was MCEnderDragon's use case.
Re-texturing of a model based on BlockState. As manually creating a ton of jsons IS annoying.
However, if we have a data generator for doing that, Then that doesn't need to exist in the our blockstate anymore.

Everything else was added on by Fry/others and is one of the reasons I think the system has gotten to complex.

**Comment by williewillus** [comment]
the biggest thing I need personally is custom model loading.


another is that the bakedmodels need to be forge wrapped VanillaModelWrappers so I can transform them with TRSRTransforms/rebake them properly. The stock vanilla ModelBlock unbaked models don't bake properly if they're builtin/generated.

**Comment by raoulvdberge** [comment]
I like the plan that @gigaherz proposed. Can this be considered? 

Also, I'm wondering what the overhead of the data generator system is if you really have a lot of permutations? Loading a bunch of small files from disk isn't exactly fast. What happens if we have a ton of mods doing this?

I'm not seeing how data generators fix the texture overrides problem in a good way. Surely having a bazillion MC model json files generated just for a simple texture override can't be a good idea?

**Comment by gigaherz** [comment]
That plan probably won't be feasible. After further work on this, I reached the conclusion that keeping the existing system and doing a "minimal workaround" may not possible, due to how so much of the code relies on assumptions that break when the system is connected with forge's modifications. 
I'm still looking at possibilities though, but don't keep your hopes too high, as there's a distinct chance that we need to give up the entire forge blockstates system right now, and not "at a later date".

**Comment by raoulvdberge** [comment]
@gigaherz  Would it be possible to create a new blockstate loader? A very simple one, that would just handle the texture overrides.

**Comment by gigaherz** [comment]
The blockstate loader isn't the problem. Both the blockstate loading and model loading work perfectly fine.
The issue is that due to the way the vanilla code is structured, I need to use the ModelLoaderRegistry to get the custom models loader to work, including making vanilla models support retexturing.
However if I make things load through the ModelLoaderRegistry, then the vanilla code path that loads the "parent": x of a model, will also use the ModelLoaderRegistry, which breaks the vanilla models.
As I said, I'm considering other approaches. I'll update this once I find a solution.

**Comment by gigaherz** [comment]
Okay I think at this point this could get some real testing and review.

There's one question that needs to be answered: Do we WANT this? Do we want to have the forge blockstates system for the life of 1.14.x in a working state?

The alternative is to fully deprecate the forge blockstates system, tell everyone "the forge blockstates system will not handle retexturing or custom models anymore, and will be removed in 1.15", and mark everything as deprecated.

My personal opinion is that having this merged would allow modders to have working custom models for their mods, and we don't have to worry about having a new model system working within weeks instead of months.

**Comment by gigaherz** [comment]
For reference, I have successfully tested the code to work for a custom model loader which indirectly loads an OBJ model.

screenshot

**Comment by williewillus** [comment]
btw, does this work properly with my specialmodels hook from #5667/#6010 ?

**Comment by gigaherz** [comment]
I will test but it should, since I didn't change anything related to that.

File: src/main/java/net/minecraftforge/client/model/BlockStateLoader.java [hunk 1]
```diff
@@ -98,10 +94,21 @@ public static BlockModelDefinition load(Reader reader, ResourceLocation location
...
-                            if (var.isVanillaCompatible())
-                                mcVars.add(new Variant(var.getModel(), (ModelRotation)var.getState().orElse(ModelRotation.X0_Y0), uvLock, weight));
-                            else
-                                mcVars.add(new ForgeVariant(location, var.getModel(), var.getState().orElse(TRSRTransformation.identity()), uvLock, var.getSmooth(), var.getGui3d(), weight, var.getTextures(), var.getOnlyPartsVariant(), var.getCustomData()));
+                            ResourceLocation modelLocation = var.getModel();
+                            if (!var.isVanillaCompatible() && bakery != null)
+                            {
+                                modelLocation = new ResourceLocation(
+                                        "internal",
+                                        String.format("%d", internalGeneratedModelId++) + "/" + location.getNamespace() + "/" + location.getPath() + "/" +
```
Comment (tterrag1098) [inline src/main/java/net/minecraftforge/client/model/BlockStateLoader.java]: just do `String.valueOf` or even `"" +`

**Review by tterrag1098 — APPROVED** [review]
Aside from one minor nitpick, this LTGM. Obviously there is a lot of cruft left lying around, but some of that will need breaking changes or new systems to remove. As a first pass to restore functionality, this is good. :+1:

File: src/main/java/net/minecraftforge/client/model/BlockStateLoader.java (hunk 1 above)
Comment (gigaherz) [reply src/main/java/net/minecraftforge/client/model/BlockStateLoader.java]: Ah forgot String.valueOf. Yes that's better.

**Pull request merged**